src/gamescope/wayland/manager.rs

```diff
@@ -0,0 +1,265 @@
+        let socket_path = format!("{}/gamescope-0", runtime_dir);
+
+        let stream = UnixStream::connect(&socket_path)?;
+        let conn = wayland_client::Connection::from_socket(stream)?;
+
+        log::info!("Connected to wayland display on: {}", socket_path);
+
+        // Retrieve the WlDisplay Wayland object from the connection. This object is
+        // the starting point of any Wayland program, from which all other objects will
+        // be created.
+        let display = conn.display();
+
+        log::debug!("Got wayland display: {:?}", display);
+
+        // Create an event queue for our event processing
+        let mut event_queue = conn.new_event_queue();
+        // An get its handle to associated new objects to it
+        let qh = event_queue.handle();
+
+        // Create a wl_registry object by sending the wl_display.get_registry request
+        // This method takes two arguments: a handle to the queue the newly created
+        // wl_registry will be assigned to, and the user-data that should be associated
+        // with this registry (here it is () as we don't need user-data).
+        let _registry = display.get_registry(&qh, ());
+
+        // Create state for the application
+        let mut state = WaylandState::new();
+
+        // To actually receive the events, we invoke the `sync_roundtrip` method. This method
+        // is special and you will generally only invoke it during the setup of your program:
+        // it will block until the server has received and processed all the messages you've
+        // sent up to now.
+        //
+        // In our case, that means it'll block until the server has received our
+        // wl_display.get_registry request, and as a reaction has sent us a batch of
+        // wl_registry.global events.
+        //
+        // `sync_roundtrip` will then empty the internal buffer of the queue it has been invoked
+        // on, and thus invoke our `Dispatch` implementation that prints the list of advertized
+        // globals.
+        event_queue.roundtrip(&mut state)?;
+
+        // Get initial Wayland result to assign control and input manager
+        let result = event_queue.blocking_dispatch(&mut state)?;
+        log::debug!("Initial wayland result: {result}");
+
+        // Run loop to listen for commands
+        tokio::task::spawn(async move {
+            while let Some(message) = command_rx.recv().await {
+                log::debug!("Wayland Message: {:?}", message);
+
+                let res: Result<(), Box<dyn Error>> = {
+                    match message.clone() {
+                        WaylandMessage::CommandTakeScreenshot(tx, file_path, screenshot_type) => {
+                            let res = Self::use_state(&mut state, |state| {
+                                log::info!("Taking screenshot of type:{screenshot_type:?} and saving to {file_path}");
+
+                                state.control.as_ref().unwrap().take_screenshot(
```

Comment:
Are we sure this unwrap is safe?

Reply:
Yea, in `use_state` we check if state is none and return an error if it is